components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2041,34 +1954,42 @@ where
 
         // No matter whether the source peer has applied to the required index,
         // it's a race to write apply state in both source delegate and target
-        // delegate. So asking the source delegate to stop first.
-        if self.ready_source_region_id != source_region_id {
-            if self.ready_source_region_id != 0 {
-                panic!(
-                    "{} unexpected ready source region {}, expecting {}",
-                    self.tag, self.ready_source_region_id, source_region_id
-                );
+        // fsm. So asking the source delegate to stop first.
+        info!(
+            "asking fsm to stop";
+            "region_id" => self.region_id(),
+            "peer_id" => self.id(),
+            "source_region_id" => source_region_id
+        );
+        fail_point!("before_handle_catch_up_logs_for_merge");
+        // Sends message to the source peer fsm and pause `exec_commit_merge` process
+        let (callback, f) = paired_future_callback::<u64>();
+        let msg = SignificantMsg::CatchUpLogs(CatchUpLogs {
+            target_region_id: self.region_id(),
+            merge: merge.to_owned(),
+            callback,
+        });
+        ctx.notifier
+            .notify_one(source_region_id, PeerMsg::SignificantMsg(msg));
+
+        // We must notify other fsm to avoid them block when this fsm is waiting for source region
+        if !ctx.flush_notifier.is_empty() || !ctx.cbs.is_empty() {
+            ctx.flush();
+            ctx.prepare_for(self);
```

Comment:
Why use `prepare_for` here?
I think we should find a proper method(i.e. not error-prone) to handle `ApplyContext` between context switch.